events/events.go

```diff
@@ -0,0 +1,142 @@
+package events
+
+import (
+	"github.com/spacemeshos/go-spacemesh/log"
+	"github.com/spacemeshos/go-spacemesh/types"
+)
+
+const (
+	NewBlock channelId = 1 + iota
+	BlockValid
+	NewAtx
+	AtxValid
+	NewTx
+	TxValid
+	RewardReceived
+)
+
+// publisher is the event publisher singleton.
+var publisher *EventPublisher
+
+// Publish publishes an event on the pubsub singleton.
+func Publish(event Event) {
+	if publisher != nil {
+		err := publisher.PublishEvent(event)
+		log.Error("pubsub error: %v", err)
+	}
+}
+
+// InitializeEventPubsub initializes the global pubsub broadcaster server
+func InitializeEventPubsub(ur string) {
+	var err error
+	publisher, err = NewEventPublisher(ur)
+	if err != nil {
+		log.Panic("cannot init pubsub: %v", err)
+	}
+}
+
+// EventPublisher is the struct that publishes events to subscribers by topics.
+type EventPublisher struct {
+	Publisher
+}
+
+// Event defines the interface that each message sent by the EventPublisher needs to implemet for it to correctly
+// be routed by topic.
+type Event interface {
+	// getChannel returns the channel on which this message will be published.
+	getChannel() channelId
+}
+
+// NewEventPublisher is a constructor for the event publisher, it received a url string in format of tcp://localhost:56565 to start
+// listening for connections.
+func NewEventPublisher(eventUrl string) (*EventPublisher, error) {
+	p, err := newPublisher(eventUrl)
+	if err != nil {
+		return nil, err
+	}
+	return &EventPublisher{*p}, nil
```

Comment:
it seems like you're creating a copy of p by dereferencing it?, it's because EventPublisher needs a value-type ? maybe it actually needs a pointer type.